websearch/websearch/src/durability.rs

```diff
@@ -0,0 +1,522 @@
+                            let empty_result = durability.persist(NoInput, Ok(Vec::new()))?;
+                            (Ok(empty_result), None)
+                        } else {
+                            // Reconstruct session for the current page
+                            let session = Impl::unwrapped_search_session(original_params.clone())?;
+                            let mut last_results = Vec::new();
+                            for _ in 0..*current_page {
+                                last_results = session.next_page()?;
+                            }
+                            // Check if this is the last page (no more results)
+                            if last_results.is_empty() {
+                                *finished = true;
+                            }
+                            let persisted_result =
+                                durability.persist(NoInput, Ok(last_results.clone()))?;
+                            (Ok(persisted_result), None)
+                        }
+                    }
+                    None => {
+                        unreachable!()
+                    }
+                };
+
+                if let Some(session) = new_live_session {
+                    *state = Some(DurableSearchSessionState::Live { session });
+                }
+
+                result
+            } else {
+                let result = durability.replay::<Vec<SearchResult>, SearchError>()?;
+                let mut state = self.state.borrow_mut();
+
+                match &mut *state {
+                    Some(DurableSearchSessionState::Live { .. }) => {
+                        unreachable!("Durable search session cannot be in live mode during replay");
+                    }
+                    Some(DurableSearchSessionState::Replay { current_page, .. }) => {
+                        *current_page += 1;
+                        Ok(result)
+                    }
+                    None => {
+                        unreachable!();
+                    }
+                }
+            }
+        }
+
+        fn get_metadata(&self) -> Option<SearchMetadata> {
+            let state = self.state.borrow();
+            match &*state {
+                Some(DurableSearchSessionState::Live { session }) => {
+                    // Always delegate to the underlying live session
+                    with_persistence_level(PersistenceLevel::PersistNothing, || {
+                        session.get_metadata()
+                    })
+                }
+                Some(DurableSearchSessionState::Replay {
+                    original_params,
+                    current_page,
+                    ..
+                }) => {
+                    // Get metadata from the current replay state and update current_page
+                    let session = Impl::unwrapped_search_session(original_params.clone()).ok()?;
+                    let mut metadata = None;
+                    for _ in 0..*current_page {
```

Comment:
This is not correct

* Either get_metadata is a pure function, in which case doing this in a loop does nothing.
* Or if it performs side effects this is incorrect as switching from replay to live mode should not execute any side effects